### PR TITLE
Update checking for existing dotenv usage

### LINKED ISSUE #11449: 9.3.2 requires .env file
# Bug report

## Describe the bug
since v.`9.3.2` and PR #10525 on `yarn dev` you'll see an error, that the loader was not able to find the `.env` file on the root layer even with disabled/not set experimental flag.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. `yarn dev`
2. check console output
3. find on 3rd position the error
## Expected behavior

no error if experimental flag is not set and no `.env` file exists on the project root

## Screenshots

```
$ next
[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
> Error:  Failed to load env from .env { Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (fs.js:498:3)
    at tryReadSync (fs.js:332:20)
    at Object.readFileSync (fs.js:369:19)
    at Object.loadEnvConfig (/foo/myapp/node_modules/next/dist/lib/load-env-config.js:9:81)
    at new Server (/foo/myapp/node_modules/next/dist/next-server/server/next-server.js:49:39)
    at new DevServer (/foo/myapp/node_modules/next/dist/server/next-dev-server.js:1:2851)
    at createServer (/foo/myapp/node_modules/next/dist/server/next.js:2:105)
    at start (/foo/myapp/node_modules/next/dist/server/lib/start-server.js:1:323)
    at nextDev (/foo/myapp/node_modules/next/dist/cli/next-dev.js:21:279)
    at commands.(anonymous function).then.exec (/foo/myapp/node_modules/next/dist/bin/next:28:339) errno: -21, syscall: 'read', code: 'EISDIR' }
```

## System information

- OS: macOS 10.15.3
- Browser: does not matter
- Version of Next.js: 9.3.2
## Additional context

none

### PR DESCRIPTION
**Pull request by ijjk**
This updates our check for `dotenv` already being installed on a project to also check higher up `package.json`s in case the app is inside of a monorepo. It currently checks higher up by looking for either a `yarn.lock` or `lerna.json` and then reading the `package.json` at the same depth. I also added tests to ensure this opts-out the `dotenv` usage correctly with this structure

Fixes: https://github.com/zeit/next.js/issues/11449

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| buildDuration | 9.8s | 9.9s | ⚠️ +108ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +1.11 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| index.html gzip | 916 B | 916 B | ✓ |
| link.html gzip | 925 B | 925 B | ✓ |
| withRouter.html gzip | 914 B | 914 B | ✓ |
| Overall change | 2.75 kB | 2.75 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| buildDuration | 10.5s | 10.4s | -29ms |
| nodeModulesSize | 52.8 MB | 52.8 MB | ⚠️ +1.11 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.24 kB | 6.24 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.2 kB | 56.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _error.js gzip | 294 kB | 294 kB | -311 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 957 B | 957 B | ✓ |
| index.js gzip | 294 kB | 294 kB | ⚠️ +555 B |
| link.js gzip | 301 kB | 302 kB | ⚠️ +602 B |
| routerDirect.js gzip | 300 kB | 300 kB | -121 B |
| withRouter.js gzip | 300 kB | 300 kB | -151 B |
| Overall change | 1.49 MB | 1.49 MB | ⚠️ +574 B |

</details>

</details>

**Comment by zanettin** [comment]
thanks a lot for the quick fix!

File: packages/next/lib/load-env-config.ts [hunk 1]
```diff
@@ -1,28 +1,46 @@
...
-      ...devDependencies,
-    })
-
-    if (allPackages.some(pkg => pkg === 'dotenv')) {
+    // check main `package.json` first
+    if (packageJsonHasDep(packageJson, 'dotenv')) {
       return false
     }
+    // check for a yarn.lock or lerna.json file in case it's a monorepo
+    const monorepoFile = findUp.sync(['yarn.lock', 'lerna.json'], { cwd: dir })
```
Comment (Timer) [inline packages/next/lib/load-env-config.ts]: `package-lock.json`, too!

Comment (ijjk) [reply packages/next/lib/load-env-config.ts]: Added

**Review by timneutkens — APPROVED** [review]

**Review by Timer — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| buildDuration | 9.8s | 10s | ⚠️ +213ms |
| nodeModulesSize | 47.7 MB | 47.7 MB | ⚠️ +1.44 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.25 kB | 6.25 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.3 kB | 56.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| index.html gzip | 920 B | 920 B | ✓ |
| link.html gzip | 928 B | 928 B | ✓ |
| withRouter.html gzip | 916 B | 916 B | ✓ |
| Overall change | 2.76 kB | 2.76 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| buildDuration | 10.5s | 10.7s | ⚠️ +239ms |
| nodeModulesSize | 47.7 MB | 47.7 MB | ⚠️ +1.44 kB |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.25 kB | 6.25 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..c6c1.js gzip | 10.1 kB | 10.1 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.3 kB | 56.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.78 kB | 4.78 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.71 kB | 6.71 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.4 kB | 51.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.js gzip | 1.24 kB | 1.24 kB | ✓ |
| _error.js gzip | 3.15 kB | 3.15 kB | ✓ |
| hooks.js gzip | 664 B | 664 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.03 kB | 2.03 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 7.86 kB | 7.86 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _app.module.js gzip | 594 B | 594 B | ✓ |
| _error.module.js gzip | 2.08 kB | 2.08 kB | ✓ |
| hooks.module.js gzip | 370 B | 370 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 1.48 kB | 1.48 kB | ✓ |
| routerDirect..dule.js gzip | 271 B | 271 B | ✓ |
| withRouter.m..dule.js gzip | 270 B | 270 B | ✓ |
| Overall change | 5.28 kB | 5.28 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | ijjk/next.js add/NEXT-153 | Change |
| - | - | - | - |
| _error.js gzip | 233 kB | 232 kB | -88 B |
| 404.html gzip | 1.32 kB | 1.32 kB | ✓ |
| hooks.html gzip | 960 B | 960 B | ✓ |
| index.js gzip | 232 kB | 232 kB | ⚠️ +250 B |
| link.js gzip | 242 kB | 243 kB | ⚠️ +84 B |
| routerDirect.js gzip | 241 kB | 241 kB | -36 B |
| withRouter.js gzip | 241 kB | 240 kB | -103 B |
| Overall change | 1.19 MB | 1.19 MB | ⚠️ +107 B |

</details>

</details>

**Pull request merged**